README.md
# Sugar Tensor - A slim tensorflow wrapper that provides syntactic sugar for tensor variables
Sugar Tensor aims to help deep learning researchers/practitioners. It adds some syntactic sugar functions to tensorflow to avoid tedious repetitive tasks. Sugar Tensor was developed under the following principles:

Current Version : __***1.0.0.2***__

## Principles
1. Don't mess up tensorflow. We provide no wrapping classes. Instead, we use a tensor itself so that developers can program freely as before with tensorflow. 
1. Don't mess up the python style.  We believe python source codes should look pretty and simple. Practical deep learning codes are very different from those of complex GUI programs. Do we really need inheritance 
and/or encapsulation in our deep learning code? Instead, we seek for simplicity and readability. For that, 
we use pure python functions only and avoid class style conventions.     

## Installation

1. Requirements
    1. tensorflow == __***1.0.0***__
    
2. Dependencies ( Will be installed automatically )
    1. tqdm >= 4.8.4

3. Installation

python 2

<pre><code>pip install --upgrade sugartensor</code></pre>

python 3

<pre><code>pip3 install sugartensor</code></pre>

docker installation : See docker [README.md](docker/README.md).

## API Document

See [SugarTensor's official API documentation]( https://buriburisuri.github.io/sugartensor/ ).

## Quick start

###Imports

<pre><code>import sugartensor as tf   # no need of 'import tensorflow'</code></pre>

## Features

### Sugar functions

All tensors--variables, operations, and constants--automatically have sugar functions which start with 'sg_' 
to avoid name space chaos. :-)   

### Chainable object syntax
  
Inspired by prettytensor library, we support chainable object syntax for all sugar functions. This should improve productivity and readability. Look at the following snippet.

<pre><code>
logit = (tf.placeholder(tf.float32, shape=(BATCH_SIZE, DATA_SIZE))
         .sg_dense(dim=400, act='relu', bn=True)
         .sg_dense(dim=200, act='relu', bn=True)
         .sg_dense(dim=10))
</code></pre>
 
### All returned objects are tensors.

In the above snippet, all values returned by sugar functions are pure tensorflow's tensor variables/constants. So, the following example is completely legal.

<pre><code>
ph = tf.placeholder(tf.float32, shape=(BATCH_SIZE, DATA_SIZE)   # <-- this is a tensor 
ph = ph.sg_dense(dim=400, act='relu', bn=True)   # <-- this is a tensor
ph = ph * 100 + 10  # <-- this is ok.
ph = tf.reshape(ph, (-1, 20, 20, 1)).conv(dim=30)   # <-- all tensorflow's function can be applied and chained.
</code></pre>

### Practical DRY (Don't repeat yourself) functions for deep learning researchers
  
We provide pre-defined powerful training and report functions for practical developers.
The following code is a full mnist training module with saver, report and early stopping support.

<pre><code>
# -*- coding: utf-8 -*-
import sugartensor as tf

# MNIST input tensor ( with QueueRunner )
data = tf.sg_data.Mnist()

# inputs
x = data.train.image
y = data.train.label

# create training graph
logit = (x.sg_flatten()
         .sg_dense(dim=400, act='relu', bn=True)
         .sg_dense(dim=200, act='relu', bn=True)
         .sg_dense(dim=10))

# cross entropy loss with logit ( for training set )
loss = logit.sg_ce(target=y)

# accuracy evaluation ( for validation set )
acc = (logit.sg_reuse(input=data.valid.image).sg_softmax()
       .sg_accuracy(target=data.valid.label, name='val'))

# train
tf.sg_train(loss=loss, eval_metric=[acc], ep_size=data.train.num_batch)
</code></pre>

You can check all statistics through the tensorboard's web interface like the following.

<p align="center">
  <img src="https://raw.githubusercontent.com/buriburisuri/sugartensor/master/sugartensor/png/image1.png" width="350"/>
</p>  
<p align="center">
  <img src="https://raw.githubusercontent.com/buriburisuri/sugartensor/master/sugartensor/png/image2.png" width="350"/>
</p>  
<p align="center">
  <img src="https://raw.githubusercontent.com/buriburisuri/sugartensor/master/sugartensor/png/image3.png" width="350"/>
</p>  

If you want to write another more complex training module without repeating saver, report, or whatever, you can do that like the following.
<pre><code>
# def alternate training func
@tf.sg_train_func   # <-- sugar annotator for training function wrapping
def alt_train(sess, opt):
    l_disc = sess.run([loss_disc, train_disc])[0]  # training discriminator
    l_gen = sess.run([loss_gen, train_gen])[0]  # training generator
    return np.mean(l_disc) + np.mean(l_gen)
    
# do training
alt_train(log_interval=10, ep_size=data.train.num_batch, early_stop=False, save_dir='asset/train/gan')    
</code></pre>

Please see the example codes in the 'sugartensor/example/' directory.

### Custom layers

You can add your own custom layer functions like the following code snippet.

```
# residual block
@tf.sg_sugar_func
def sg_res_block(tensor, opt):
    # default rate
    opt += tf.sg_opt(size=3, rate=1, causal=False)

    # input dimension
    in_dim = tensor.get_shape().as_list()[-1]

    # reduce dimension
    input_ = (tensor
              .sg_bypass(act='relu', bn=(not opt.causal), ln=opt.causal)
              .sg_conv1d(size=1, dim=in_dim/2, act='relu', bn=(not opt.causal), ln=opt.causal))

    # 1xk conv dilated
    out = input_.sg_aconv1d(size=opt.size, rate=opt.rate, causal=opt.causal, act='relu', bn=(not opt.causal), ln=opt.causal)

    # dimension recover and residual connection
    out = out.sg_conv1d(size=1, dim=in_dim) + tensor

    return out

# inject residual block
tf.sg_inject_func(sg_res_block)
```

For more information, see [ByteNet example code](https://github.com/buriburisuri/ByteNet/blob/master/train.py) 
or [WaveNet example code](https://github.com/buriburisuri/speech-to-text-wavenet/blob/master/train.py).  


### Multip GPU support

You can train your model with multiple GPUs using sg_parallel decorator as follow:

```
# batch size
batch_size = 128


# MNIST input tensor ( batch size should be adjusted for multiple GPUS )
data = tf.sg_data.Mnist(batch_size=batch_size * tf.sg_gpus())

# split inputs for each GPU tower
inputs = tf.split(data.train.image, tf.sg_gpus(), axis=0)
labels = tf.split(data.train.label, tf.sg_gpus(), axis=0)


# simple wrapping function with decorator for parallel training
@tf.sg_parallel
def get_loss(opt):

    # conv layers
    with tf.sg_context(name='convs', act='relu', bn=True):
        conv = (opt.input[opt.gpu_index]
                .sg_conv(dim=16, name='conv1')
                .sg_pool()
                .sg_conv(dim=32, name='conv2')
                .sg_pool()
                .sg_conv(dim=32, name='conv3')
                .sg_pool())

    # fc layers
    with tf.sg_context(name='fcs', act='relu', bn=True):
        logit = (conv
                 .sg_flatten()
                 .sg_dense(dim=256, name='fc1')
                 .sg_dense(dim=10, act='linear', bn=False, name='fc2'))

        # cross entropy loss with logit
        return logit.sg_ce(target=opt.target[opt.gpu_index])

# parallel training ( same as single GPU training )
tf.sg_train(loss=get_loss(input=inputs, target=labels), ep_size=data.train.num_batch)

```

# Author

Namju Kim (namju.kim@kakaocorp.com) at KakaoBrain Corp.

CHANGELOG.md
## 1.0.0.2 ( 2017-03-20 )

Bugfixes:
    - fix a running mean and variance update rule in batch normal 
    - fix a gru bug

Features :
    - support multiple GPU support
    - add regularizer related functions  
    - add summary on/off context

## 1.0.0.1 ( 2017-02-28 )

Bugfixes:
    - fix a missed tensorflow 1.0 refactoring at sg_upconv1d()
    - fix a missed tensorflow 1.0 refactoring at sg_ctc_loss()
    
Features :
    - support VALID padding at sg_upconv() and sg_upconv1d()
    - add zero_out parameter at sg_rnn(), sg_gru() and sg_lstm() 

## 1.0.0.0 ( 2017-02-22 )

Features :
    - scope support at batch normalization

Refactored :
    - adapted to tensorflow 1.0.0
    
     


sugartensor/sg_net.py
from __future__ import absolute_import
import sugartensor as tf


__author__ = 'namju.kim@kakaocorp.com'


#
# VGG nets
#


@tf.sg_sugar_func
def sg_vgg_16(tensor, opt):
    r"""Applies vgg 16 model.
    
    Note that the fc layers in the original architecture 
      will be replaced with fully convolutional layers.
      For convenience, We still call them fc layers, though.
    
    Args:
      tensor: A `Tensor`
      opt:
        num_class: An integer. Number of class. Default is 1000.
        conv_only: Boolean. If True, fc layers are not applied. Default is False.
        squeeze: Boolean. If True (default), the dimensions with size 1 in the final outputs will be removed.
        act: String. 'relu' (default). the activation function name
        bn: True or False(default). If True, batch normal will be applied
        reuse: Boolean(Optional). If True, all variables will be loaded from previous network.
        name: String(Optional). If provided, used as the scope name of this network
    
    Returns:
      A `Tensor`.
    """
    opt += tf.sg_opt(num_class=1000, conv_only=False, squeeze=True, act='relu', bn=False)

    # convolution layers
    with tf.sg_context(name=opt.name, act=opt.act, bn=opt.bn, reuse=opt.reuse):
        conv = (tensor
                .sg_conv(dim=64, name='conv1_1')
                .sg_conv(name='conv1_2')
                .sg_pool()
                .sg_conv(dim=128, name='conv2_1')
                .sg_conv(name='conv2_2')
                .sg_pool()
                .sg_conv(dim=256, name='conv3_1')
                .sg_conv(name='conv3_2')
                .sg_conv(name='conv3_3')
                .sg_pool()
                .sg_conv(dim=512, name='conv4_1')
                .sg_conv(name='conv4_2')
                .sg_conv(name='conv4_3')
                .sg_pool()
                .sg_conv(dim=512, name='conv5_1')
                .sg_conv(name='conv5_2')
                .sg_conv(name='conv5_3')
                .sg_pool())

    # fully convolution layers
    with tf.sg_context(name=opt.name, act=opt.act, pad='VALID', bn=opt.bn, reuse=opt.reuse):
        fc = (conv
              .sg_conv(dim=4096, size=7, dout=opt.dout, name='fc1')
              .sg_conv(dim=4096, size=1, dout=opt.dout, name='fc2')
              .sg_conv(dim=opt.num_class, size=1, act='linear', name='fc3'))

    if opt.conv_only:
        return conv
    else:
        if opt.squeeze:
            return fc.sg_squeeze(axis=(1, 2))
        else:
            return fc


@tf.sg_sugar_func
def sg_vgg_19(tensor, opt):
    r"""Applies vgg 19 model.
    
    Note that the fc layers in the original architecture 
      will be replaced with fully convolutional layers.
      For convenience, We still call them fc layers, though.
    
    Args:
      tensor: A `Tensor`.
      opt:
        num_class: An integer. Number of class. Default is 1000.
        conv_only: Boolean. If True, fc layers are not applied. Default is False.
        squeeze: Boolean. If True (default), the dimensions with size 1 in the final outputs will be removed.
        act: String. 'relu' (default). the activation function name
        bn: True or False(default). If True, batch normal will be applied
        reuse: Boolean(Optional). If True, all variables will be loaded from previous network.
        name: String(Optional). If provided, used as the scope name of this network
    
    Returns:
      A `Tensor`.
    """
    opt += tf.sg_opt(num_class=1000, conv_only=False, squeeze=True, act='relu')

    # convolution layers
    with tf.sg_context(name=opt.name, act=opt.act, bn=opt.bn, reuse=opt.reuse):
        conv = (tensor
                .sg_conv(dim=64, name='conv1_1')
                .sg_conv(name='conv1_2')
                .sg_pool()
                .sg_conv(dim=128, name='conv2_1')
                .sg_conv(name='conv2_2')
                .sg_pool()
                .sg_conv(dim=256, name='conv3_1')
                .sg_conv(name='conv3_2')
                .sg_conv(name='conv3_3')
                .sg_conv(name='conv3_4')
                .sg_pool()
                .sg_conv(dim=512, name='conv4_1')
                .sg_conv(name='conv4_2')
                .sg_conv(name='conv4_3')
                .sg_conv(name='conv4_4')
                .sg_pool()
                .sg_conv(dim=512, name='conv5_1')
                .sg_conv(name='conv5_2')
                .sg_conv(name='conv5_3')
                .sg_conv(name='conv5_4')
                .sg_pool())

    # fully convolution layers
    with tf.sg_context(name=opt.name, act=opt.act, pad='VALID', bn=opt.bn, reuse=opt.reuse):
        fc = (conv
              .sg_conv(dim=4096, size=7, dout=opt.dout, name='fc1')
              .sg_conv(dim=4096, size=1, dout=opt.dout, name='fc2')
              .sg_conv(dim=opt.num_class, size=1, act='linear', name='fc3'))

    if opt.conv_only:
        return conv
    else:
        if opt.squeeze:
            return fc.sg_squeeze(axis=(1, 2))
        else:
            return fc


#
# Residual nets ( version 2 )
#


@tf.sg_sugar_func
def sg_resnet_50(x, opt):
    r"""Applies residual net 50 model.
    
    Note that the fc layers in the original architecture 
      will be replaced with fully convolutional layers.
      For convenience, We still call them fc layers, though.
    
    Args:
      x: A `Tensor`.
      opt:
        num_class: An integer. Number of class. Default is 1000.
        conv_only: Boolean. If True, fc layers are not applied. Default is False.
        squeeze: Boolean. If True (default), the dimensions with size 1 in the final outputs will be removed.
        act: String. 'relu' (default). the activation function name
        reuse: Boolean(Optional). If True, all variables will be loaded from previous network.
        name: String(Optional). If provided, used as the scope name of this network
    
    Returns:
      A `Tensor`. 
    """
    return _resnet_graph(x, opt, [3, 4, 6, 3])


@tf.sg_sugar_func
def sg_resnet_101(x, opt):
    r"""Applies residual net 101 model.
    
    Note that the fc layers in the original architecture 
      will be replaced with fully convolutional layers.
      For convenience, We still call them fc layers, though.
    
    Args:
      x: A `Tensor`.
      opt:
        num_class: An integer. Number of class. Default is 1000.
        conv_only: Boolean. If True, fc layers are not applied. Default is False.
        squeeze: Boolean. If True (default), the dimensions with size 1 in the final outputs will be removed.
        act: String. 'relu' (default). the activation function name
        reuse: Boolean(Optional). If True, all variables will be loaded from previous network.
        name: String(Optional). If provided, used as the scope name of this network

    Returns:
      A `Tensor`. 
    """
    return _resnet_graph(x, opt, [3, 4, 23, 3])


@tf.sg_sugar_func
def sg_resnet_152(x, opt):
    r"""Applies residual net 152 model.
    
    Note that the fc layers in the original architecture 
      will be replaced with fully convolutional layers.
      For convenience, We still call them fc layers, though.
    
    Args:
      x: A `Tensor`.
      opt:
        num_class: An integer. Number of class. Default is 1000.
        conv_only: Boolean. If True, fc layers are not applied. Default is False.
        squeeze: Boolean. If True (default), the dimensions with size 1 in the final outputs will be removed.
        act: String. 'relu' (default). the activation function name
        reuse: Boolean(Optional). If True, all variables will be loaded from previous network.
        name: String(Optional). If provided, used as the scope name of this network

    Returns:
      A `Tensor`. 
    """
    return _resnet_graph(x, opt, [3, 8, 36, 3])


@tf.sg_sugar_func
def sg_resnet_200(x, opt):
    r"""Applies residual net 200 model.
    
    Note that the fc layers in the original architecture 
      will be replaced with fully convolutional layers.
      For convenience, We still call them fc layers, though.
    
    Args:
      x: A `Tensor`.
      opt:
        num_class: An integer. Number of class. Default is 1000.
        conv_only: Boolean. If True, fc layers are not applied. Default is False.
        squeeze: Boolean. If True (default), the dimensions with size 1 in the final outputs will be removed.
        act: String. 'relu' (default). the activation function name
        reuse: Boolean(Optional). If True, all variables will be loaded from previous network.
        name: String(Optional). If provided, used as the scope name of this network

    Returns:
      A `Tensor`. 
    """
    return _resnet_graph(x, opt, [3, 24, 36, 3])


@tf.sg_sugar_func
def sg_densenet_121(x, opt):
    r"""Applies dense net 121 model.

    Note that the fc layers in the original architecture
      will be replaced with fully convolutional layers.
      For convenience, We still call them fc layers, though.

    Args:
      x: A `Tensor`.
      opt:
        k: An integer. The Growth rate of densenet. Default is 32.
        num_class: An integer. Number of class. Default is 1000.
        conv_only: Boolean. If True, fc layers are not applied. Default is False.
        squeeze: Boolean. If True (default), the dimensions with size 1 in the final outputs will be removed.
        act: String. 'relu' (default). the activation function name
        reuse: Boolean(Optional). If True, all variables will be loaded from previous network.
        name: String(Optional). If provided, used as the scope name of this network

    Returns:
      A `Tensor`.
    """
    return _densenet_graph(x, opt + tf.sg_opt(k=32), [6, 12, 24, 16])


@tf.sg_sugar_func
def sg_densenet_169(x, opt):
    r"""Applies dense net 169 model.

    Note that the fc layers in the original architecture
      will be replaced with fully convolutional layers.
      For convenience, We still call them fc layers, though.

    Args:
      x: A `Tensor`.
      opt:
        k: An integer. The Growth rate of densenet. Default is 32.
        num_class: An integer. Number of class. Default is 1000.
        conv_only: Boolean. If True, fc layers are not applied. Default is False.
        squeeze: Boolean. If True (default), the dimensions with size 1 in the final outputs will be removed.
        act: String. 'relu' (default). the activation function name
        reuse: Boolean(Optional). If True, all variables will be loaded from previous network.
        name: String(Optional). If provided, used as the scope name of this network

    Returns:
      A `Tensor`.
    """
    return _densenet_graph(x, opt + tf.sg_opt(k=32), [6, 12, 32, 32])


@tf.sg_sugar_func
def sg_densenet_201(x, opt):
    r"""Applies dense net 201 model.

    Note that the fc layers in the original architecture
      will be replaced with fully convolutional layers.
      For convenience, We still call them fc layers, though.

    Args:
      x: A `Tensor`.
      opt:
        k: An integer. The Growth rate of densenet. Default is 32.
        num_class: An integer. Number of class. Default is 1000.
        conv_only: Boolean. If True, fc layers are not applied. Default is False.
        squeeze: Boolean. If True (default), the dimensions with size 1 in the final outputs will be removed.
        act: String. 'relu' (default). the activation function name
        reuse: Boolean(Optional). If True, all variables will be loaded from previous network.
        name: String(Optional). If provided, used as the scope name of this network

    Returns:
      A `Tensor`.
    """
    return _densenet_graph(x, opt + tf.sg_opt(k=32), [6, 12, 48, 32])


@tf.sg_sugar_func
def sg_densenet_161(x, opt):
    r"""Applies dense net 161 model.

    Note that the fc layers in the original architecture
      will be replaced with fully convolutional layers.
      For convenience, We still call them fc layers, though.

    Args:
      x: A `Tensor`.
      opt:
        k: An integer. The Growth rate of densenet. Default is 48.
        num_class: An integer. Number of class. Default is 1000.
        conv_only: Boolean. If True, fc layers are not applied. Default is False.
        squeeze: Boolean. If True (default), the dimensions with size 1 in the final outputs will be removed.
        act: String. 'relu' (default). the activation function name
        reuse: Boolean(Optional). If True, all variables will be loaded from previous network.
        name: String(Optional). If provided, used as the scope name of this network

    Returns:
      A `Tensor`.
    """
    return _densenet_graph(x, opt + tf.sg_opt(k=48), [6, 12, 36, 24])


#
# Residual net utility
#

@tf.sg_sugar_func
def sg_resnet_layer(x, opt):
    r"""Applies basic architecture of residual net.
    
    Note that the fc layers in the original architecture 
      will be replaced with fully convolutional layers.
      For convenience, We still call them fc layers, though.
    
    Args:
      x: A `Tensor`.
      opt:
          dim: An integer. Dimension for this resnet layer
          num: Number of times to repeat
          act: String. 'relu' (default). the activation function name
          reuse: Boolean(Optional). If True, all variables will be loaded from previous network.
          name: String. (optional) Used as convolution layer prefix

    Returns:
      A `Tensor`. 
    """
    assert opt.dim is not None, 'dim is mandatory.'
    assert opt.num is not None, 'num is mandatory.'

    # default stride
    opt += tf.sg_opt(stride=1, act='relu')

    # format convolutional layer name
    def cname(index):
        return opt.name if opt.name is None else opt.name + '_%d' % index

    with tf.sg_context(bias=False, reuse=opt.reuse):
        # 1st block
        out = (x
               .sg_bypass(act=opt.act, bn=True, name=cname(0))
               .sg_conv(dim=opt.dim, size=1, act=opt.act, stride=opt.stride, bn=True, name=cname(1))
               .sg_conv(dim=opt.dim, size=3, act=opt.act, bn=True, name=cname(2))
               .sg_conv(dim=opt.dim*4, size=1, name=cname(3)))
        out += x.sg_conv(dim=opt.dim * 4, size=1, stride=opt.stride, name=cname(4))

        # middle blocks
        for i in range(1, opt.num):
            out_new = (out
                       .sg_bypass(act=opt.act, name=cname(4 * i + 1))
                       .sg_conv(dim=opt.dim, size=1, act=opt.act, bn=True, name=cname(4 * i + 2))
                       .sg_conv(dim=opt.dim, size=3, act=opt.act, bn=True, name=cname(4 * i + 3))
                       .sg_conv(dim=opt.dim*4, size=1, name=cname(4 * i + 4)))
            out += out_new

    return out


# construct residual network graphs
def _resnet_graph(x, opt, nums):

    # default option
    opt += tf.sg_opt(num_class=1000, conv_only=False, squeeze=True, act='relu')

    # convolution layers ( residual net v2 arch )
    with tf.sg_context(name=opt.name):
        conv = (x
                .sg_conv(dim=64, size=7, stride=2, bias=False, reuse=opt.reuse, name='conv1')
                .sg_pool(size=3, stride=2, pad='SAME')
                .sg_resnet_layer(dim=64, num=nums[0], stride=1, act=opt.act, reuse=opt.reuse, name='conv2')
                .sg_resnet_layer(dim=128, num=nums[1], stride=2, act=opt.act, reuse=opt.reuse, name='conv3')
                .sg_resnet_layer(dim=256, num=nums[2], stride=2, act=opt.act, reuse=opt.reuse, name='conv4')
                .sg_resnet_layer(dim=512, num=nums[3], stride=2, act=opt.act, reuse=opt.reuse, name='conv5')
                .sg_bypass(act=opt.act, bn=True, reuse=opt.reuse, name='final_act')  # final activation
                .sg_pool(size=7, stride=1, avg=True))  # global average pool

        # fully convolution layers
        fc = conv.sg_conv(dim=opt.num_class, size=1, act='linear', bn=False, reuse=opt.reuse, name='fc')

    # return selectively
    if opt.conv_only:
        return conv
    else:
        if opt.squeeze:
            return fc.sg_squeeze(axis=(1, 2))
        else:
            return fc


#
# Dense net utility
#

@tf.sg_sugar_func
def sg_densenet_layer(x, opt):
    r"""Applies basic architecture of densenet layer.

    Note that the fc layers in the original architecture
      will be replaced with fully convolutional layers.
      For convenience, We still call them fc layers, though.

    Args:
      x: A `Tensor`.
      opt:
          dim: An integer. Dimension for this resnet layer
          num: Number of times to repeat
          act: String. 'relu' (default). the activation function name
          trans: Boolean. If True(default), transition layer will be applied.
          reuse: Boolean(Optional). If True, all variables will be loaded from previous network.
          name: String. (optional) Used as convolution layer prefix

    Returns:
      A `Tensor`.
    """
    assert opt.dim is not None, 'dim is mandatory.'
    assert opt.num is not None, 'num is mandatory.'

    # default stride
    opt += tf.sg_opt(stride=1, act='relu', trans=True)

    # format convolutional layer name
    def cname(index):
        return opt.name if opt.name is None else opt.name + '_%d' % index

    # dense layer
    with tf.sg_context(bias=False, reuse=opt.reuse):
        out = x
        for i in range(opt.num):
            # dense block
            out_new = (out
                       .sg_bypass(act=opt.act, bn=True, name=cname(3 * i + 1))
                       .sg_conv(dim=opt.dim // 4, size=1, act=opt.act, bn=True, name=cname(3 * i + 2))
                       .sg_conv(dim=opt.dim, size=3, name=cname(3 * i + 3)))
            out = tf.concat([out_new, out], 3)

        # transition layer
        if opt.trans:
            out = (out
                   .sg_bypass(act=opt.act, bn=True, name=cname(3 * i + 4))
                   .sg_conv(size=1, name=cname(3 * i + 5))
                   .sg_pool(avg=True))

    return out


# construct dense network graphs
def _densenet_graph(x, opt, nums):

    # default option
    opt += tf.sg_opt(num_class=1000, k=32, conv_only=False, squeeze=True, act='relu')

    # convolution layers ( dense net arch )
    with tf.sg_context(name=opt.name):
        conv = (x
                .sg_conv(dim=opt.k, size=7, stride=2, bias=False, reuse=opt.reuse, name='conv1')
                .sg_pool(size=3, stride=2, pad='SAME')
                .sg_densenet_layer(dim=opt.k, num=nums[0], act=opt.act, reuse=opt.reuse, name='conv2')
                .sg_densenet_layer(dim=opt.k, num=nums[1], act=opt.act, reuse=opt.reuse, name='conv3')
                .sg_densenet_layer(dim=opt.k, num=nums[2], act=opt.act, reuse=opt.reuse, name='conv4')
                .sg_densenet_layer(dim=opt.k, num=nums[3], act=opt.act, trans=False, reuse=opt.reuse, name='conv5')
                .sg_bypass(act=opt.act, bn=True, reuse=opt.reuse, name='final_act')  # final activation
                .sg_pool(size=7, stride=1, avg=True))  # global average pool

        # fully convolution layers
        fc = conv.sg_conv(dim=opt.num_class, size=1, act='linear', bn=False, reuse=opt.reuse, name='fc')

    # return selectively
    if opt.conv_only:
        return conv
    else:
        if opt.squeeze:
            return fc.sg_squeeze(axis=(1, 2))
        else:
            return fc
